ObserverSet.swift

```diff
@@ -26,122 +26,198 @@
+open class ObserverSet<Parameters> {
+	
+	// MARK: - Private properties
+	
+	private var entries: [ObserverSetEntry<Parameters>] = []
+	
+	// MARK: - Initialisers
+	
+	/**
+	Creates a new instance of an observer set.
+	
+	- returns: A new instance of an observer set.
+	*/
+	public init() {}
+	
+	// MARK: - Public functions
+	
+	/**
+	Adds a new observer to the set.
+	
+	- parameter queue: The queue to call the function on when the notification is delivered. If nil (the default) it will be delivered on the same queue the notification was sent.
+	- parameter observer: The object that is to be notified.
+	- parameter callBack: The function to call on the observer when the notification is to be delivered.
+	
+	- returns: An entry in the list of observers, which can be used later to remove the observer.
+	*/
+	@discardableResult
+	open func add<ObserverType: AnyObject>(queue: DispatchQueue? = nil, _ observer: ObserverType, _ callBack: @escaping (ObserverType) -> (Parameters) -> Void) -> ObserverSetEntry<Parameters> {
+		let entry = ObserverSetEntryImpl<ObserverType, Parameters>(queue: queue, observer: observer, callBack: callBack)
+		synchronized {
+			self.entries.append(entry)
+		}
+		return entry
+	}
+	
+	/**
+	Adds a new function (or closure) to call when a notification is to be delivered.
+	This callback will be notified until it is removed by passing the return value to the `remove(_:)` method
+	
+	The return value is not discardable as the callback should be removed manually. To attach it to the life cycle of an object, use the `add` method that takes a `lifeCycleObserver` parameter.
+	
+	- parameter queue: The queue to call the function on when the notification is delivered. If nil (the default) it will be delivered on the same queue the notification was sent.
+	- parameter callBack: The function to call when the notification is to be delivered.
+	
+	- returns: An entry in the list of observers, which should be used later to remove the observer.
+	*/
+	open func add(queue: DispatchQueue? = nil, _ callBack: @escaping (Parameters) -> Void) -> ObserverSetEntry<Parameters> {
+		add(queue: queue, self) { _ in callBack }
+	}
+	
+	/**
+	Adds a new function (or closure) to call when a notification is to be delivered.
+	The callback will be removed when the `lifeCycleObserver` object is deallocated
+	
+	- parameter queue: The queue to call the function on when the notification is delivered. If nil (the default) it will be delivered on the same queue the notification was sent.
+	- parameter callBack: The function to call when the notification is to be delivered.
+	
+	- returns: An entry in the list of observers, which can be used later to remove the observer.
+	*/
+	@discardableResult
+	open func add<ObserverType: AnyObject>(queue: DispatchQueue? = nil, _ lifeCycleObserver: ObserverType, _ callBack: @escaping (Parameters) -> Void) -> ObserverSetEntry<Parameters> {
```

Comment:
Added this function for convenience when you want to use a closure but want to tie to the lifecycle of an object.

One use case is instead of this:

```
let someObserverSet = ObserverSet<Int>()
let someObserverSet = ObserverSet<String>()

class Foo {
  private let observer1: Any
  private let observer2: Any
  init() {
      // Use closures to ignore the parameters that are different
      observer1 = someObserverSet.add { [weak self] _ in self?.update() }
      observer2 = anotherObserverSet.add { [weak self] _ in self?.update() }
  }
  func update() {
  
  }
  deinit {
    someObserverSet.remove(observer1)
    someObserverSet.remove(observer2)
  }
}
```

Just do this:
```
class Foo {
  init() {
      // Use closures to ignore the parameters that are different
      someObserverSet.add(self) { [weak self] _ in self?.update() }
      anotherObserverSet.add(self) { [weak self] _ in self?.update() }
  }
  func update() {
  
  }
}
```

An alternative would be to allow observers that have no parameters for any ObserverSet that has a non void Parameters type, but that would remove some of the type safety.

```
class Foo {
  init() {
      // Use closures to ignore the parameters that are different
      someObserverSet.add(self, Foo.update)
      anotherObserverSet.add(self, Foo.update) // oops I meant to do Foo.bar which takes the parameter but there is no error because any () -> Void function can be used as a callback for any ObserverSet
  }
  func update() {
  
  }

  func bar(_: String) { }
}
```